### ISSUE TITLE
Some ideas about detecting ss flow

### ISSUE THREAD

**Issue by bosskwei** [shadowsocks/shadowsocks-org#86]
Hi All,
These days my connections to vps with shadowsocks seems not stable, I think gfw could detect my vps, so I make some simple attempts as follows:

figure_1 (dataset_1): normal tcp (95% http / https) flow captured by wireshark (about 1GB)
figure_2 (dataset_2): ss flow filtered from dataset_1 (about 50MB)

left chart: data len (amplitude) in time domain
mid chart: Fourier domain analyse (Meaningless value in axis-x)
right chart: histogram of data len

![figure_1](https://user-images.githubusercontent.com/12762888/31449103-a00afd00-aed8-11e7-85fc-51aeecb7d1fd.png)
![figure_2](https://user-images.githubusercontent.com/12762888/31449105-a0b870ca-aed8-11e7-9e58-1db48839c8fb.png)

In the two figures, we could easily noticed that ss-server works similarly to a web server, but all payload are encrypted.

And here's a method may could detect ss:
```
if similar_to_web_server(ip, batch_amplitude):
    if all_data_are_encrypted(ip, batch_payload):
        block_ip(ip)
```

In addition, whether it's neural network or decision trees, training on encrypted ss payload will learn nothing ( white noise). But if use whole flow (included ip header, tcp / udp header), network could extract features from addr/ port dst, addr / port src, flow_len, as mentioned above.

**Here's the demo code**
[https://github.com/bosskwei/temp/blob/master/detect_shadowsocks.py](https://github.com/bosskwei/temp/blob/master/detect_shadowsocks.py)

**You should change line 65 to pcap file path, and 45 to your local_ip such as 192.168.0.1**

**Comment by arthurkiller** [comment]
I come up with same issues. Clarify what actually happening is much more important.
Let me tell what's going on:

I have 4 independent port listened on my VPS to provide service.
First day, One of them was blocked. But others are reachable. I still can SSH to remote login my host.

Next day , all my ports was blocked including 22 port. I can not even remote login my host anymore.

My service on AWS got some trouble these days too. Traffic from Peking out to Singapore looks good, but incoming traffic was shit! 60% packet got lost and I DONT KNOW WHY.

So , In my view, there is no evidence shown that there comes any tech-way to detect SS trafic. What happened is only CAUSE OF THE 19th MEETING OF CCP

**Comment by arthurkiller** [comment]
@bosskwei 
Do we need t make SS repack the tcp packet in to several random length ss-packet?

**Comment by tony1016** [comment]
Means if I use simple-obfs http mode,then will pass the gfw because all my data seem like unencrypted??

**Comment by Mygod** [comment]
https://github.com/shadowsocks/shadowsocks-org/issues/81#issuecomment-336019891

**Comment by bosskwei** [comment]
@tony1016 Actually I don't know how to deal with it, obfs may be lead to other problems...

**Comment by ZhangJiupeng** [comment]
@tony1016 Not pretty sure. I tried this method yesterday, to pretend my traffic as HTTP flows, it works soon, but my port is surprisingly banned today (maybe the GFW is just simply triggered by detecting large traffic and then block all the suspected traffic), and I am still figuring out what has happened. To locate these problems, we still need more reports on GFW's recent actions.

**Comment by Mygod** [comment]
I encourage everyone to do experiments and post their results online. It's helpful for analyzing GFW via crowdsourcing in some sense.

**Comment by bosskwei** [comment]
@Mygod Hi, I upload a demo code to verify this idea, help me test it..

**Comment by leonshaw** [comment]
I just wonder what is the criteria you use to conclude two traffic pattern are "similar". Have you tried other types of traffic?
Generally ss doesn't change packet length pattern (much). The reason it's like HTTP is probably that the payload is HTTP traffic.
To verify if the firewall detect ss in this manner, we need some sort of controlled experiment.

**Comment by bosskwei** [comment]
@leonshaw 
1. See my code,
```
square_loss_freq = np.abs(freq_feature - F)
square_loss_hist = np.abs(hist_feature - H)
```
Other traffic such as IM, is totally different from http, example:
![figure_3](https://user-images.githubusercontent.com/12762888/31538773-2b09f15c-b039-11e7-9b04-2ce95164bc76.png)
And, my work is just a demo, I tried to extract features manual. I do not have enough time and dataset to train a neural network.
Whether it's neural network or decision trees, training on encrypted ss payload will learn nothing ( white noise). But if use whole flow (included ip header, tcp / udp header), network could extract features as above.

2. Yes, (1) traffic like HTTP, (2) but no tls handshake, no http header is enough to confirm your server as a unknown proxy.

3. So, I put my idea and code here, you could do experiment and paste your results.

**Comment by Mygod** [comment]
Please try to refrain from doing statistical/timing analysis. So far I've only seen evidence of GFW not doing them.

**Comment by bosskwei** [comment]
Closed because there is no evidence that gfw doing them.

**Comment by klzgrad** [comment]
You can keep the ideas open. Basically nobody has no evidence about nothing and we're all grasping at air so there's no harm in trying even if it ends up not getting implemented. But one thing I'm sure of is GFW will not use this alone in identifying ss traffic, if it does at all, that is, the alleged compute difficulty of running classifiers on all traffic isn't really a concern.

I was made aware of the fact that GFW used netflows to classify network-level host behaviors several years ago. The claim was it was able to identify VPN servers by netflows alone, which I didn't believe but it now makes complete sense as a way to isolate traffic of suspect hosts for finer grained traffic analysis. Picture the traffic patterns of a ss host - a single or a few clients with very long session time and large traffic size and ratio, which would be a fairly unique profile for classification but also general enough for any kind of proxy servers. So it's quite possible there's already a relatively small subset of hosts identified for further analysis from netflow analysis which renders the compute problem irrelevant, and GFW has been doing this for quite some years.

The random forest paper by Deng et al. has poor presentation but its premise is legit, as in the features it proposes to use are well-established in the traffic analysis literature to be informative. People have used it to identify URLs in TLS traffic among other things. The point of doing this is to gain finer understanding of host behaviors at application level so the false positive rate is lowered with the extra information. Now, the question worth asking is how much information is in the features (netflows, packet size distributions) and how much discriminative power it can produce.

Also keep in mind that even if the paper's quality is poor, they were not doing this for shits and giggles. This is the tip of an iceberg and there will be a bigger project with allocated budget and deeper plans running behind. And the experimental/amateurish nature of this research means it would not be deployed to production so soon nevertheless the symptoms we've seen so far simply mean other similar projects are going on elsewhere in parallel. The primary goals for these projects would be usually to identify botnets and malware hosts as an order of magnitude more of a threat than shadowsocks.

**Comment by bosskwei** [comment]
@klzgrad Thank you for your advice, but I will still keep this issue close unless gfw could indeed detect ss.
And if anyone has ideas about it please reference this issue.

**Comment by klzgrad** [comment]
(Using Chrome's network stack as example)

Suppose you use Shadowsocks to access an HTTPS site. SSL handshakes ensue inside the Shadowsocks tunnel.

Common SSL *packet* sizes (note that SSL records are being sent with TCP_NODELAY):

* The first client side packet:

183 + p(SNI): Unpadded ClientHello + SNI (a normal distribution e.g. ~[195, 212], 183 is the lower bound with empty SNI)
211 + p(SNI): Unpadded ClientHello + SNI + Session ID(32) - channel_id(4) (211 is the lower bound with empty SNI)
517: Padded ClientHello

* The next one or two client side packets:

51: ChangeCipherSpec (5 + 1) + Finished (5 + 40)
93: ClientKeyExchange (5 + 37, ECDH 32) + ChangeCipherSpec (5 + 1) + Finished (5 + 40)
126: ClientKeyExchange (5 + 70, ECDH 65) + ChangeCipherSpec (5 + 1) + Finished (5 + 40)
216: ChangeCipherSpec (5 + 1) + EncryptedHandshakeMessage (5 + 160) + Finished (5 + 40)
258: ClientKeyExchange (5 + 37, ECDH 32) + ChangeCipherSpec (5 + 1) + EncryptedHandshakeMessage (5 + 160) + Finished (5 + 40)
318: ClientKeyExchange (5 + 262, RSA 256) + ChangeCipherSpec (5 + 1) + Finished (5 + 40)
324: ClientKeyExchange (5 + 262, RSA 256) + ChangeCipherSpec (5 + 1) + Finished (5 + 64)

* If ALPN is h2 (this is also possible to distinguish from the length of "http/1.1" in ServerHello), the next client side packet will be (also TCP_NODELAY):

93: TLS header (5) + TLS GCM Nonce (8) + TLS GCM MAC (16) + [h2 Magic + SETTINGS + WINDOW_UPDATE] (64)

---

Now let's look at Shadowsocks's AEAD scheme:

```
[encrypted payload length][length tag][encrypted payload][payload tag]
```
"Tag Size" is 16 in all ciphers. Thus there is a fixed overhead of 34 (2 + 16 + 16). The "encrypted payload" will be the SSL records counted above.

There are not many combinations of ways to segment the initial SSL handshake records into one or two packets. Add the overhead of 34 to the above packet sizes and there you have it. I can provide a toy example:
```
handshake_sizes = [51, 93, 126, 216, 258, 318, 324]
if sizes[0] - 34 == 517 and sizes[1] - 34 in handshake_sizes and sizes[2] - 34 == 93:
    print('Detected Shadowsocks tunnel with SSL content (padded ClientHello, alpn=h2)')
```

**Comment by bosskwei** [comment]
@klzgrad yes, detect ssl-in-shadowsocks is easy as mentioned above, random padding is really needed.

**Comment by shinku721** [comment]
@bosskwei I think one more thing you need to test with is other encrypted VPN/proxy protocols. I guess they are similar to shadowsocks, too. However, a protocol similar to nothing may have the greatest risk.
